driverdataaggregation/main.py

```diff
@@ -0,0 +1,609 @@
+                                 producer app
+        :type clients_per_node: list[int], optional
+        :param client_nodes: number of compute nodes to use for the aggregation
+                             producer app
+        :type client_nodes: list[int], optional
+        :param iterations: number of append/retrieve loops run by the applications
+        :type iterations: int
+        :param tensor_bytes: list of tensor sizes in bytes
+        :type tensor_bytes: list[int], optional
+        :param tensors_per_dataset: list of number of tensors per dataset
+        :type tensor_per_dataset: list[int], optional
+        :param client_threads: list of the number of client threads used for data
+                               aggregation
+        :type client_threads: list[int], optional
+        :param cpu_hyperthreads: the number of hyperthreads per cpu.  This is done
+                                 to request that the consumer application utilizes
+                                 all physical cores for each client thread.
+        :type cpu_hyperthreads: int, optional
+        """
+        logger.info("Starting dataset aggregation scaling tests")
+        logger.info(f"Running with database backend: {get_db_backend()}")
+        logger.info(f"Running with launcher: {launcher}")
+
+        exp = create_folder(exp_name, launcher)
+
+        for db_node_count in db_nodes:
+
+            # start the database only once per value in db_nodes so all permutations
+            # are executed with the same database size without bringin down the database
+            db = start_database(exp,
+                                db_node_feature,
+                                db_port,
+                                db_node_count,
+                                db_cpus,
+                                None, # not setting threads per queue in throughput tests
+                                net_ifname,
+                                run_db_as_batch,
+                                db_hosts)
+
+
+            for c_nodes, cpn, _bytes, t_per_dataset, c_threads in product(
+                client_nodes, clients_per_node, tensor_bytes, tensors_per_dataset, client_threads
+            ):
+                logger.info(f"Running with threads: {c_threads}")
+                # setup a an instance of the C++ driver and start it
+                aggregation_producer_sessions = \
+                    self._create_aggregation_producer_session_cpp(exp, node_feature, c_nodes, cpn,
+                                                        db_node_count,
+                                                        db_cpus, iterations,
+                                                        _bytes, t_per_dataset)
+
+                # setup a an instance of the C++ driver and start it
+                aggregation_consumer_sessions = \
+                    self._create_aggregation_consumer_session_cpp(exp, node_feature, c_nodes, cpn,
```

Comment:
Okay, kinda dangerous problem I ran into here!

So when running this study with 60 producer nodes each with 48 clients, I was running into the following error
```
Error from shell: srun: Warning: can't honor --ntasks-per-node set to 1 which doesn't match the requested tasks 2880 with the number of requested nodes 1. Ignoring --ntasks-per-node.
srun: error: Unable to create step for job 3729164: More processors requested than permitted
```
This appears to be because this is the actual srun call that is being made (unimportant stuff replaced with `...`)
```
env SSDB=... \
  /opt/slurm/20.11.5/bin/srun ... \
    --ntasks=2880 ... --nodes=1 ... 1024000 4
#   ^^^^^^^^^^^^^ <---^^^^^^^^^---------- Number of tasks looks good (48 * 60 = 2880)
#                     ^^^^^^^^^ <---------- but trying to put them all on one node :(
```

This is because `node_feature` is a mutable data structure, and because smartsim doesn't actually make a deep copy of the run args dictionary when calling `Experiment.create_run_settings`, this call to this fn will actually mutated the run settings of the producer model. This is weird and gross and should not work this way imo (I submitted a ticket to address this) but for now this is the way that it does.

Luckily the solution to fix this is to simply make a deep copy of `node_feature` pass the original to `create_aggregation_producer_session_cpp` and the copy to `create_aggregation_consumer_session_cpp` or vice-versa.

Reply:
Seems like this issue also exists for the python and fs versions of this test, but it looks like the solution should be pretty similar